alpa/pipeline_parallel/pipeshard_executable.py

```diff
@@ -90,12 +122,83 @@ def __init__(self,
-        self._instantiate_nccl_groups(pipeshard_config.device_str_groups)
+        #self._instantiate_nccl_groups(pipeshard_config.device_str_groups)
+
         self.resharding_tasks = pipeshard_config.resharding_tasks
+
+        for resharding_task in self.resharding_tasks:
+            if global_config.resharding_mode == "send_recv":
+                task_dones = []
+                for v_worker, task in resharding_task.sender_tasks.items():
+                    uuid = resharding_task.send_worker_task_ids[v_worker]
+                    worker = resharding_task.collective_group.mesh_workers[v_worker.index]
+                    task_dones.append(
+                            worker.put_resharding_send_task.remote(
+                                uuid, task, resharding_task.collective_group.group_name))
+                for v_worker, task in resharding_task.receiver_tasks.items():
+                    uuid = resharding_task.recv_worker_task_ids[v_worker]
+                    worker = resharding_task.collective_group.mesh_workers[v_worker.index]
+                    task_dones.append(
+                            worker.put_resharding_recv_task.remote(
+                                uuid, task, resharding_task.collective_group.group_name))
+                ray.get(task_dones)
+
+                task_dones = []
+                if resharding_task.is_local_allgather_task:
+                    uuid = resharding_task.allgather_uuid
+                    task_spec = resharding_task.task_spec
+                    hlo = compile_allgather(task_spec.aval.shape, task_spec.aval.dtype,
+                                                task_spec.dst_sharding_spec,
+                                                task_spec.final_dst_spec,
+                                                np.prod(resharding_task.dst_mesh.shape))
+                    for v_worker in resharding_task.dst_mesh.workers:
+                        worker = resharding_task.collective_group.mesh_workers[v_worker.index]
+                        task_dones.append(
+                                    worker.put_executable.remote(uuid, UtilMeshWorkerExecutable,
+                                                                 hlo))
+                ray.get(task_dones)
+            else:
+                task_dones = []
+                for v_worker, task in resharding_task._broadcast_tasks.items():
+                    uuid = resharding_task.broadcast_worker_task_ids[v_worker]
+                    worker = resharding_task.collective_group.mesh_workers[v_worker.index]
+                    task_dones.append(
+                    worker.put_resharding_broadcast_task.remote(
+                            uuid, task, resharding_task.collective_group.group_name))
+                ray.get(task_dones)
+
+        ##### For cross-mesh resharding #####
+        if not self.v_meshes.launched_nccl:
+            self._instantiate_nccl_groups(pipeshard_config.device_str_groups)
```

Comment:
as mentioned above, we always instantiate or inherit from an existing physical mesh group before this step, so we just use the original code.